File: BotNet/client/CrazyUncleBurton/java/Player.java
import com.sun.jna.Pointer;

///Stores information about a player in the game
class Player
{
  Pointer ptr;
  int ID;
  int iteration;
  public Player(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.playerGetId(ptr);
    iteration = BaseAI.iteration;
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.players.length; i++)
    {
      if(BaseAI.players[i].ID == ID)
      {
        ptr = BaseAI.players[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///Allows a player to display a message to the screen.
  int talk(String message)
  {
    validify();
    return Client.INSTANCE.playerTalk(ptr, message);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.playerGetId(ptr);
  }
  ///Player's Name
  public String getPlayerName()
  {
    validify();
    return Client.INSTANCE.playerGetPlayerName(ptr);
  }
  ///Player's points, one with more at end of game wins
  public int getByteDollars()
  {
    validify();
    return Client.INSTANCE.playerGetByteDollars(ptr);
  }
  ///Player's machine cycles, used to create new Virus's
  public int getCycles()
  {
    validify();
    return Client.INSTANCE.playerGetCycles(ptr);
  }
  ///The amount of time this player has before timing out
  public int getTime()
  {
    validify();
    return Client.INSTANCE.playerGetTime(ptr);
  }

}

File: BotNet/client/CrazyUncleBurton/c/Player.cpp
// -*-c++-*-

#include "Player.h"
#include "game.h"


Player::Player(_Player* pointer)
{
    ptr = (void*) pointer;
}

int Player::id()
{
  return ((_Player*)ptr)->id;
}

char* Player::playerName()
{
  return ((_Player*)ptr)->playerName;
}

int Player::byteDollars()
{
  return ((_Player*)ptr)->byteDollars;
}

int Player::cycles()
{
  return ((_Player*)ptr)->cycles;
}

int Player::time()
{
  return ((_Player*)ptr)->time;
}


int Player::talk(char* message)
{
  return playerTalk( (_Player*)ptr, message);
}



std::ostream& operator<<(std::ostream& stream,Player ob)
{
  stream << "id: " << ((_Player*)ob.ptr)->id  <<'\n';
  stream << "playerName: " << ((_Player*)ob.ptr)->playerName  <<'\n';
  stream << "byteDollars: " << ((_Player*)ob.ptr)->byteDollars  <<'\n';
  stream << "cycles: " << ((_Player*)ob.ptr)->cycles  <<'\n';
  stream << "time: " << ((_Player*)ob.ptr)->time  <<'\n';
  return stream;
}
